### PR TITLE
ROX-19562: Add annotations to rate limit managed-central routes traffic

### PR DESCRIPTION
**Pull request by rhybrillou**
## Description

In order to protect against (D)DOS attacks, the OpenShift routes to Central managed by FleetShard sync are being annotated in order to activate traffic throttling features.

The default values were derived from the haproxy metrics available in the current ACS CS instances.
- The most busy central shows a total HTTP response count of 9 averaged over a 5 minute period.
Assuming all the averaged requests were a burst of concurrent requests, these could have been about 3000 requests.
Adding a safety 10 factor, and using powers of two as throttling value, the default would be 32768 concurrent requests.
- The most busy central reported up to 43 concurrent connections on a 5 minute period. The limit derived from here is 512 concurrent connections, allowing at most the half as being established at the same time.

## Checklist (Definition of Done)

- [ ] Unit and integration tests added
- [ ] Added test description under `Test manual`
- [ ] Documentation added if necessary (i.e. changes to dev setup, test execution, ...)
- [ ] CI and all relevant tests are passing
- [x] Add the ticket number to the PR title if available, i.e. `ROX-12345: ...`
- [ ] Discussed security and business related topics privately. Will move any security and business related topics that arise to private communication channel.
- [ ] Add secret to app-interface Vault or Secrets Manager if necessary
- [ ] RDS changes were e2e tested [manually](../docs/development/howto-e2e-test-rds.md)
- [ ] Check AWS limits are reasonable for changes provisioning new resources

## Test manual

**TODO:** Add manual testing efforts

```
# To run tests locally run:
make db/teardown db/setup db/migrate
make ocm/setup OCM_OFFLINE_TOKEN=<ocm-offline-token> OCM_ENV=development
make verify lint binary test test/integration
```

### CODE REVIEW

File: fleetshard/config/config.go [hunk 1]
```diff
@@ -64,12 +65,20 @@ type Telemetry struct {
...
 type SecretEncryption struct {
 	Type  string `env:"SECRET_ENCRYPTION_TYPE" envDefault:"local"`
 	KeyID string `env:"SECRET_ENCRYPTION_KEY_ID"`
 }
 
+// RouteConfig defines parameters to configure routes.
+type RouteConfig struct {
+	ThrottlingEnabled bool `env:"ROUTE_ENABLE_THROTTLING" envDefault:"true"`
+	ConcurrentTCP     int  `env:"ROUTE_CONCURRENT_TCP_CONNECTIONS" envDefault:"512"`
+	RateHTTP          int  `env:"ROUTE_HTTP_REQUEST_RATE" envDefault:"32768"`
```
Comment (janisz) [inline fleetshard/config/config.go:78]: Add queries that gives this numbers

File: fleetshard/pkg/k8s/route.go [hunk 2]
```diff
@@ -126,15 +204,35 @@ func (s *RouteService) UpdateReencryptRoute(ctx context.Context, route *openshif
 	return nil
 }
 
+func (s *RouteService) configurePassthroughRoute(_ context.Context, route *openshiftRouteV1.Route, remoteCentral private.ManagedCentral) (*openshiftRouteV1.Route, error) {
```
[the remark this answers is not in the document]
Comment (rhybrillou) [reply fleetshard/pkg/k8s/route.go:207]: This function is expected to be of type `configureRouteFunc`, which is the common denominator for configuration of passthrough and reencrypt routes. Re-encrypt needs the context to fetch data from k8s, so the context has to be part of the function signature.

[the remark this answers is not in the document]
Comment (rhybrillou) [reply fleetshard/pkg/k8s/route.go:207]: This function should be of type `configureRouteFunc`, like `configureReencryptRoute` which uses the context.

File: fleetshard/pkg/k8s/route_test.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
+package k8s
```
[the remark this answers is not in the document]
Comment (rhybrillou) [reply fleetshard/pkg/k8s/route_test.go]: #1382

File: fleetshard/pkg/k8s/route_test.go [hunk 4]
```diff
@@ -65,16 +99,26 @@ var (
...
 )
 
 var (
 	passThroughRouteExtractors = map[string]routeFieldExtractor{
-		"name":                          extractRouteObjectMetaName,
-		"namespace":                     extractRouteObjectMetaNamespace,
-		"\"managed-by\" label":          extractRouteManagedByLabel,
+		"name":                                  extractRouteObjectMetaName,
+		"namespace":                             extractRouteObjectMetaNamespace,
+		"\"managed-by\" label":                  extractRouteManagedByLabel,
```
Comment (janisz) [inline fleetshard/pkg/k8s/route_test.go:117]: nit
```suggestion
		`"managed-by" label`:                  extractRouteManagedByLabel,
```

File: fleetshard/pkg/k8s/route_test.go [hunk 5]
```diff
@@ -210,34 +258,42 @@ var (
...
+		"\"timeout\" annotation":                               extractRouteReEncryptTimeoutAnnotation,
+		"spec host":                                            extractRouteSpecHost,
+		"spec target port type":                                extractRouteSpecTargetPortType,
+		"spec target port string value":                        extractRouteSpecTargetPortStrVal,
+		"spec \"to\" kind":                                     extractRouteSpecToKind,
+		"spec \"to\" name":                                     extractRouteSpecToName,
+		"spec TLS termination":                                 extractRouteSpecTLSTermination,
+		"spec TLS destination CA certificate":                  extractRouteSpecTLSDestinationCACertificate,
+		"spec TLS key":                                         extractRouteSpecTLSKey,
+		"spec TLS certificate":                                 extractRouteSpecTLSCertificate,
```
Comment (janisz) [inline fleetshard/pkg/k8s/route_test.go:288]: consider adding a empty line and a section with new objects. Order is not important and then the diff will be smaller.

**Review by janisz — APPROVED** [review]

**Pull request merged**